### PR TITLE
WIP: break everything, new docs

### PR DESCRIPTION
**Pull request by mlubin**
Beginning of a rewrite of the docs using Documenter (#169), incorporating status changes (#164) and a new attribute concept.

We'll use this as the starting point for the documentation tutorial tomorrow.

Preview of the documentation: http://www.juliaopt.org/MathProgBase.jl/latest/

### CODE REVIEW

File: src/SolverInterface/sets.jl [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+"""
+struct Interval{T} <: AbstractSet
+    lower::T
+    upper::T
+end
+
+dimension(s::Interval) = 1
+
+"""
+    addconstraint!(m::AbstractMathProgModel, b, a_varidx, a_coef, Q_vari, Q_varj, Q_val, S::AbstractSet)
```
Comment (odow) [inline src/SolverInterface/sets.jl]: s/a_coef/a_val

File: src/SolverInterface/sets.jl [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+Add the constraint
+```math
+a^Tx + b + \\frac{1}{2}x^TQx \\in S
+```
+where ``a`` is a sparse vector specified in tuple form by
+`a_varidx`, and `a_val`; ``b`` is a scalar;
+the symmetric matrix ``Q_i`` is defined by the triplets in `Q_vari`, `Q_varj`,
```
Comment (odow) [inline src/SolverInterface/sets.jl]: s/Q_i/Q

File: src/SolverInterface/attributes.jl [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    getattribute(m, VariableResult(), ref)
+    getattribute(m, VariableResult(5), [ref1,ref2])
+    getattribute(m, OtherAttribute("something specific to cplex"))
+"""
+function getattribute end
+
+"""
+    getattribute!(output, m::AbstractMathProgModel, attr::AbstractAttribute)
+
+An in-place version of `getattribute!`. Return an attribute of the model `m` specified by attribute type `attr` into output vector `attr`.
```
Comment (odow) [inline src/SolverInterface/attributes.jl]: s/getattribute!/getattribute

File: src/SolverInterface/attributes.jl [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    VariableUpperBound
+
+Upper-bound constraints for the variables. `Inf` is valid as no bound.
+"""
+struct VariableUpperBound <: AbstractAttribute end
+
+"""
+    VariablePrimal(N)
+
+The assignment to the primal variables in result `N`. If `N` is omitted, it is interpreted as 1.
```
Comment (joehuchette) [inline src/SolverInterface/attributes.jl]: Is the plan to implement this with a ``VariablePrimal() = VariablePrimal(1)`` constructor?

Comment (mlubin) [inline src/SolverInterface/attributes.jl]: Yeah that works.

**Comment by odow** [comment]
Some more potential attributes
- [ ] `NumberVariables`
- [ ] `NumberConstraints` split by set type?
- [ ] List of constraint sets currently in the model?

File: src/SolverInterface/constraints.jl [hunk 5]
```diff
@@ -28,9 +28,14 @@ isvalid(m::AbstractMathProgModel, ref::ConstraintReference) = throw(MethodError(
...
 
-Delete the referenced variable from the model.
+Delete the referenced constraint from the model.
+
+    delete!(m::AbstractMathProgModel, refs::Vector{ConstraintReference})
+
+Delete the referenced constraints in the vector `refs` from the model.
 """
 Base.delete!(m::AbstractMathProgModel, ref::ConstraintReference) = throw(MethodError())
+Base.delete!(m::AbstractMathProgModel, refs::Vector{ConstraintReference}) = throw(MethodError())
```
Comment (odow) [inline src/SolverInterface/constraints.jl]: The fallback should probably be
`Base.delete!(m::AbstractMathProgModel, refs::Vector{ConstraintReference}) = delete!.(m, refs)`

File: src/SolverInterface/variables.jl [hunk 6]
```diff
@@ -25,10 +25,15 @@ isvalid(m::AbstractMathProgModel, ref::VariableReference) = throw(MethodError())
...
-
+    
 Delete the referenced variable from the model.
+
+    delete!(m::AbstractMathProgModel, refs::Vector{VariableReference})
+
+Delete the referenced variables in the vector `refs` from the model.
 """
 Base.delete!(m::AbstractMathProgModel, ref::VariableReference) = throw(MethodError())
+Base.delete!(m::AbstractMathProgModel, refs::Vector{VariableReference}) = throw(MethodError())
```
Comment (odow) [inline src/SolverInterface/variables.jl]: The fallback should probably be
`Base.delete!(m::AbstractMathProgModel, refs::Vector{VariableReference}) = delete!.(m, refs)`

File: src/SolverInterface/constraints.jl [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    delete!(m::AbstractMathProgModel, ref::ConstraintReference)
+
+Delete the referenced constraint from the model.
+
+    delete!(m::AbstractMathProgModel, refs::Vector{ConstraintReference})
+
+Delete the referenced constraints in the vector `refs` from the model.
+"""
+Base.delete!(m::AbstractMathProgModel, ref::ConstraintReference) = throw(MethodError())
+Base.delete!(m::AbstractMathProgModel, refs::Vector{ConstraintReference}) = throw(MethodError())
```
Comment (odow) [inline src/SolverInterface/constraints.jl]: The fallback should probably be
`Base.delete!(m::AbstractMathProgModel, refs::Vector{ConstraintReference}) = delete!.(m, refs)`

File: src/SolverInterface/variables.jl [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    delete!(m::AbstractMathProgModel, ref::VariableReference)
+    
+Delete the referenced variable from the model.
+
+    delete!(m::AbstractMathProgModel, refs::Vector{VariableReference})
+
+Delete the referenced variables in the vector `refs` from the model.
+"""
+Base.delete!(m::AbstractMathProgModel, ref::VariableReference) = throw(MethodError())
+Base.delete!(m::AbstractMathProgModel, refs::Vector{VariableReference}) = throw(MethodError())
```
Comment (odow) [inline src/SolverInterface/variables.jl]: The fallback should probably be
`Base.delete!(m::AbstractMathProgModel, refs::Vector{VariableReference}) = delete!.(m, refs)`

File: docs/src/index.md [hunk 9]
```diff
@@ -1 +1,7 @@
-# MathProgBase documentation
+# MathProgBase
+
+Solver-independent functions and low-level interfaces for Mathematical Programming
+
+[MathProgBase.jl](https://github.com/JuliaOpt/MathProgBase.jl) provides high-level one-shot functions for linear and mixed-integer programming, as well as a solver-independent low-level interface for implementing advanced techniques that require efficiently solving a sequence of linear programming problems.
```
Comment (blegat) [inline docs/src/index.md]: Isn't it a bit restricted to talk about linear programming here ?

File: src/SolverInterface/constraints.jl [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+```math
+Ax + b + q(x) \\in S
+```
+where ``A`` is a sparse matrix specified in triplet form by
+`a_constridx`, `a_varidx`, and `a_coef`; ``b`` is a vector;
+``q(x)`` is a vector with component ``(q(x))_k`` defined to be ``\\frac{1}{2}x^TQ_kx``
+where the symmetric matrix ``Q_k`` is defined by the triplets in `Q_vari`, `Q_varj`,
+`Q_coef` for which `Q_constridx` equals `k`; and the set ``S`` is defined by `S`.
+
+Duplicate indices in either ``A`` or  a ``Q`` matrix are accepted and will be summed together. Off-diagonal entries of ``Q`` will be mirrored, so either the upper triangular or lower triangular entries of ``Q`` should be provided. If entries for both ``(i,j)`` and ``(j,i)`` are provided, these are considered duplicate terms. `a_varidx`, `Q_vari`, `Q_varj` should be collections of `VariableReference` objects.
```
Comment (blegat) [inline src/SolverInterface/constraints.jl]: It is unclear what "mirrored" means. Does `2xy` corresponds to `xy + yx` or `2xy + 2yx` ?

File: docs/src/index.md [hunk 11]
```diff
@@ -1 +1,3 @@
-# MathProgBase documentation
+# MathProgBase
+
+[MathProgBase.jl](https://github.com/JuliaOpt/MathProgBase.jl) is a standarized API for Mathematical Optimization solvers. [JuMP](https://github.com/JuliaOpt/JuMP.jl) uses MathProgBase as a solver-independent low-level backend, but MathProgBase may be used freely without JuMP. In addition to the low-level API, MathProgBase provides one-shot functions for linear, mixed-integer, and quadratic optimiztion problems.
```
Comment (odow) [inline docs/src/index.md]: s/standarized/standardized
s/optimiztion/optimization

**Comment by chriscoey** [comment]
some TODOs

- [ ]  in the definitions at http://www.juliaopt.org/MathProgBase.jl/latest/objectives.html, replace indices `idx` with variable references (and maybe type the function arguments)

**Comment by odow** [comment]
More TODO
- [ ] query portions of objective and constraint by variable reference. What happens for quadratic?

File: docs/src/apireference.md [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+```
+
+## Attributes
+
+These are used to get and set properties of the model.
+
+```@docs
+AbstractAttribute
+cangetattribute
+getattribute
```
Comment (joehuchette) [inline docs/src/apireference.md]: ``getattribute!`` is missing here

File: src/SolverInterface/objectives.jl [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+triplets in `Q_vari`, `Q_varj`, `Q_coef`.
+
+Duplicate indices in either the ``a`` vector or the ``Q`` matrix are accepted and will be
+summed together. Off-diagonal entries of ``Q`` will be mirrored, so either the
+upper triangular or lower triangular entries of ``Q`` should be provided. If
+entries for both ``(i,j)`` and ``(j,i)`` are provided, these are considered
+duplicate terms. `a_varidx`, `Q_vari`, `Q_varj` should be collections of
+`VariableReference` objects.
+"""
+function setobjective! end
```
Comment (joehuchette) [inline src/SolverInterface/objectives.jl]: Worth having a method that only adds the linear part? We could have a fallback method in MPB to fill in the empty quadratic part
```jl
setobjective!(m::AbstractMathProgModel, N::Int, b, a_varidx, a_coef)
```

Comment (mlubin) [inline src/SolverInterface/objectives.jl]: Yes, that's a TODO

**Comment by mlubin** [comment]
TODO: for constraint attributes, ``cangetattribute`` should take the constraint reference

**Comment by mlubin** [comment]
I'm planning on moving this branch into a new package. Potential names: MathOptAPI (MOA) (@odow) and MathOptInterface

Given that NLP is still half-baked and doesn't play well with the new design, we can leave MPB around for NLP solvers until we have a good unified API for it.

**Comment by odow** [comment]
You're stealing my name! Although I've been leaning more to [Kōkako](https://en.wikipedia.org/wiki/K%C5%8Dkako).jl recently...

**Comment by mlubin** [comment]
Development will continue here: https://github.com/JuliaOpt/MathOptInterface.jl

**Pull request closed without merging**